Improve YAML Parsing

Trenly opened the pull request:
# Summary
This PR updates the YamlCreate.ps1 script to be more fully-featured and uses [powershell-yaml](https://www.powershellgallery.com/packages/powershell-yaml/0.4.2) ([See it on GitHub](https://github.com/cloudbase/powershell-yaml)) to handle the parsing of keys and values

## Major Updates
- Script documentation added under /doc/tools/YamlCreate.md
- PackageId, PackageVersion, Mode, and Settings parameters added as optional
- Validation of inputs against the manifest schema
- Ability to fill in PR parameters when submitting PR's automatically
  - Ability to link issues when submitting a PR
- Quick update mode added for when only InstallerURL's change
- Allows for locales other than en-US as the defaultLocale
- User can customize their settings for the script
- Checks for PR's that may be for the same change

## Minor Updates
- Bumped script version and updated header
- Installer keys are sorted based on the order in the schema
- Automatic detection of architecture and installer type where possible 
- Automatic detection of package family name, where possible
- Changed the termination types to be `exit` when a user choice exits the script and `throw` when the script terminates unexpectedly
- Throws descriptive exceptions with inner exceptions for debugging
- Delete mode for removing package versions, with reason
- Handles file names in a safer manner
- Handles branch names in a safer manner
- Uses quasi-unique branch names
- Suppresses much of the git output which is not necessary to show

-----

- [x] Have you signed the [Contributor License Agreement](https://cla.opensource.microsoft.com/microsoft/winget-pkgs)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/microsoft/winget-pkgs/pulls) for the same manifest update/change?

-----
Resolves #21940
Resolves #21219

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/winget-pkgs/pull/21204)

Trenly committed:
Install powershell-yaml

Trenly committed:
Enforce ordering of keys

Trenly committed:
Create function for adding list parameters

Trenly committed:
Write Version Manifest using Yaml Parser

Trenly committed:
Add To-Do Section

Trenly committed:
Rebase on master and cleanup commit history

Trenly committed:
Revert inadvertent changes from testing

Trenly committed:
Implement logic for installers using YAML parser

Trenly committed:
Fix statement in incorrect position

Trenly committed:
Remove RAW writing

Trenly commented:
Something weird happened with the commit history when I tried to rebase on master. I revered back to before the rebase. I'll rebase to resolve the merge conflicts before going too much further

Trenly committed:
Begin reading MultiManifests using Yaml parser

Trenly committed:
Cleanup unneeded variables

Trenly committed:
Bump Version

Trenly committed:
Read parameters from singletons

Trenly committed:
Add Product Code

Trenly committed:
Abstract ReadInstallerManifest to sub-function

Trenly committed:
Fix: Handle uninitialized variables

Trenly committed:
Merge New/Update, add EditMetadata

Trenly commented:
@OfficialEsco I believe this is all I have left -
```yaml
<#
TO-DO:
    - Handle writing null parameters as comments
    - Ensure licensing for powershell-yaml is met
#>
```
I'm not sure how to write comments with powershell-yaml, so that may be difficult. As far as I can tell, there aren't any licensing requirement for powershell-yaml since we aren't using the source code in any way, but I would like someone else to verify that.

Other than those two things, I believe this is ready for testing

Trenly committed:
Fix: Change EditMetadata to work on singleton manifests

denelon commented:
@Trenly is it [this one](https://github.com/cloudbase/powershell-yaml/blob/master/LICENSE)?

Trenly commented:
> @Trenly is it [this one](https://github.com/cloudbase/powershell-yaml/blob/master/LICENSE)?

Yes, It is that one

Trenly committed:
Fix: Update Git commit messages for accuracy

Trenly committed:
Remove Empty Manifest Folders

OfficialEsco commented on Tools/YamlCreate.ps1:
```diff
@@ -409,113 +337,65 @@ Function Read-WinGet-InstallerValues {
+
+    do {
+        if (!$InstallerSuccessCodes) { $InstallerSuccessCodes = '' }
+        $script:InstallerSuccessCodes = PromptInstallerManifestValue $InstallerSuccessCodes 'Commands' '[Optional] List of additional non-zero installer success exit codes other than known default values by winget (Max 16)'
```
```suggestion
        $script:InstallerSuccessCodes = PromptInstallerManifestValue $InstallerSuccessCodes 'InstallerSuccessCodes' '[Optional] List of additional non-zero installer success exit codes other than known default values by winget (Max 16)'
```
Guessing this is the correct part?

OfficialEsco reviewed (changes requested):
**Issue 1:**
```
[Optional] Enter any Protocols the application provides a handler for. For example: http, https (Max 16)
Protocols:

[Optional] Enter any Commands or aliases to run the application. For example: msedge (Max 16)
Commands:

[Optional] List of additional non-zero installer success exit codes other than known default values by winget (Max 16)
Commands:
```

**Feature Request:**
Keep Metadata outside of hardcoded metadata (do we have to hardcode all the different metadata fields?)

**Untested:**
Does it work as expected with additional locales? (Copying them to the new version)

I'll keep testing it if i can find any non submitted versions which have gotten harder 😠

Edit: On last issue revealed itself at the end

```
Submit PR?
Do you want to submit your PR now?
[Y] Yes  [N] No (default is 'Y'): yUpdate
PS C:\Users\Esco\Github\winget-pkgs>
```

Trenly committed:
Fix: Correct text

Trenly committed:
Fix: Remove code for testing

Trenly commented:
> **Issue 1:**
> 
> ```
> [Optional] Enter any Protocols the application provides a handler for. For example: http, https (Max 16)
> Protocols:
> 
> [Optional] Enter any Commands or aliases to run the application. For example: msedge (Max 16)
> Commands:
> 
> [Optional] List of additional non-zero installer success exit codes other than known default values by winget (Max 16)
> Commands:
> ```
>  
Fixed

-----
> **Feature Request:**
> Keep Metadata outside of hardcoded metadata (do we have to hardcode all the different metadata fields?)
> 
I think I have a few ideas how to do this. It will take some work, but shouldn't be too difficult. Basically what I need to do is if the old metadata files exist, update the fields in-place. I'll work on that over the next few days

-----
> 
> **Untested:**
> Does it work as expected with additional locales? (Copying them to the new version)

Yes, I did test this using Mozilla.Thunderbird. Creating a new version copies the locale files and updates the version. I would appreciate if others could test this also though. 

-----
> I'll keep testing it if i can find any non submitted versions which have gotten harder 😠
> 
Thanks!!

-----
> Edit: On last issue revealed itself at the end
> 
> ```
> Submit PR?
> Do you want to submit your PR now?
> [Y] Yes  [N] No (default is 'Y'): yUpdate
> PS C:\Users\Esco\Github\winget-pkgs>
> ```
> 
Fixed

OfficialEsco commented:
@quhxl

Trenly committed:
Feature: If old manifests exist, update in place to preserve extra keys

Trenly commented:
@OfficialEsco The feature you suggested was easier than I thought to implement if you would like to test it now

OfficialEsco commented:
> The feature you suggested was easier than I thought to implement if you would like to test it now

Thank you, expected it was sorta simple because of the Yaml parsing 😁 https://github.com/microsoft/winget-pkgs/pull/17873/files us a good test subject and its working perfectly!

Trenly commented:
> > The feature you suggested was easier than I thought to implement if you would like to test it now
> 
> Thank you, expected it was sorta simple because of the Yaml parsing 😁 [Plex v1.33.0](https://github.com/microsoft/winget-pkgs/pull/17873/files) is a good test subject and its working perfectly!
> 
> Uhm, not really a issue, but why does it do this?
> 
> ```
> InstallerSuccessCodes:
> - "1223"
> ```

It looks like it is handling it as a string and that is how it parses numbers as strings. I'll see if there is a way for it to not do that, but since you say its not an issue I'll keep it as low priority

Trenly committed:
Fix: Null Error when creating new

Trenly committed:
Fix: Cast only when installer code

Trenly commented:
@OfficialEsco I added a cast to Int which seems to have fixed it. Let me know what you think

Trenly commented:
@denelon Would you be able to check and remove the `Validation-Merge-Conflict` label? I believe that is an artifact from the rebase that was rolled back

OfficialEsco reviewed (changes requested):
Issue reading certain characters
```
[Optional] Enter the application Copyright. For example: Copyright (c) Microsoft Corporation
Old Variable: Copyright Â© 2005-2021 Piriform Software Ltd
```
Probably because of this, it NEEDS to be UTF-8 (Pretty sure this happened in one of the last commits)
![image](https://user-images.githubusercontent.com/15158490/126342417-d1e89002-14bd-4f40-abba-e4cd4ec3a7b4.png)

Not sure if this is your fault, but InstallerLocale prompt is being ignored and always spitting out `InstallerLocale: en-US`

Trenly committed:
Fix: Respect installer locale

Trenly commented:
> Issue reading certain characters
> 
> ```
> [Optional] Enter the application Copyright. For example: Copyright (c) Microsoft Corporation
> Old Variable: Copyright Â© 2005-2021 Piriform Software Ltd
> ```
> 
> Probably because of this, it NEEDS to be UTF-8 (Pretty sure this happened in one of the last commits)
> ![image](https://user-images.githubusercontent.com/15158490/126342417-d1e89002-14bd-4f40-abba-e4cd4ec3a7b4.png)
> 
> Not sure if this is your fault, but InstallerLocale prompt is being ignored and always spitting out `InstallerLocale: en-US`

Hmm. When I tried reading in UTF8 the parser didn't like it. Let me see if there's a workaround

Trenly committed:
Fix: Use UTF8 Encoding

OfficialEsco reviewed (changes requested):
The manifests are still saved as `UTF-16 LE` 🙄

Spotted another issue, when i add a new InstallerSuccessCodes it gets put at the bottom of the Manifest, guessing any new variable will be put there

Trenly commented:
> The manifests are still saved as `UTF-16 LE` 🙄
> 
Thanks for the thorough testing. Does it matter if it is UTF-8-BOM or UTF-8-NoBOM ?

> Spotted another issue, when i add a new InstallerSuccessCodes it gets put at the bottom of the Manifest, guessing any new variable will be put there
>
Looks like it has to do with the way I'm handling preserving keys that aren't defined in the script. I can write the file from the start, which preserves the order but discards extraneous keys; or I can preserve those keys but not preserve order. I have a few ideas though that might allow me to fix this. 

Does it matter if the keys that aren't explicitly defined in the script go to the bottom?

OfficialEsco commented:
> Thanks for the thorough testing. Does it matter if it is UTF-8-BOM or UTF-8-NoBOM ?

From experience from when i did a overhaul, yes it does, it has to be UTF-8-NoBOM

> Looks like it has to do with the way I'm handling preserving keys that aren't defined in the script. I can write the file from the start, which preserves the order but discards extraneous keys; or I can preserve those keys but not preserve order. I have a few ideas though that might allow me to fix this.
>
> Does it matter if the keys that aren't explicitly defined in the script go to the bottom?

Well, it has to do with Standardized Manifest/Metadata so its easy to look for and look at, aka so its not a complete mess.

One thing i have been thinking about is, would it be possible to take the order from the Metadata Schema?
https://raw.githubusercontent.com/microsoft/winget-cli/master/schemas/JSON/manifests/v1.0.0/manifest.installer.1.0.0.json
I actually think Winget-create does this

Trenly commented:
> > Thanks for the thorough testing. Does it matter if it is UTF-8-BOM or UTF-8-NoBOM ?
> 
> From experience from when i did a overhaul, yes it does, it has to be UTF-8-NoBOM
> 
Sounds good. I think thats as simple as putting the re-writes back in

> > Looks like it has to do with the way I'm handling preserving keys that aren't defined in the script. I can write the file from the start, which preserves the order but discards extraneous keys; or I can preserve those keys but not preserve order. I have a few ideas though that might allow me to fix this.
> Well, it has to do with Standardized Manifest/Metadata so its easy to look for and look at, aka so its not a complete mess.
> 
> One thing i have been thinking about is, would it be possible to use the Metadata Schema to get the order?
> https://raw.githubusercontent.com/microsoft/winget-cli/master/schemas/JSON/manifests/v1.0.0/manifest.installer.1.0.0.json
> I actually think Winget-create does this
>
Hmm. That's an interesting thought. I'm not sure how that would work. I don't know Powershell all that well, but it might be possible to do

OfficialEsco commented:
Huh, did a search and found this https://hochwald.net/powershell-convert-yaml-to-json-and-json-to-yaml/
Been messing around and got to this point: 
```
$SchemaRaw = (Get-Content -raw .\Downloads\test.json)
$SchemaJson = @($SchemaRaw | ConvertFrom-Json)

$SchemaJson.properties                                   # For all the possible top layer metadata
$SchemaJson.definitions.installer.properties             # For all the Installer Metadata
```
Now, if there is a way to clean that up then we have a very nice list, would also be useful for using the patterns and min/max letters instead of hard coding it into the script

Trenly commented:
> Huh, did a search and found this https://hochwald.net/powershell-convert-yaml-to-json-and-json-to-yaml/
> Been messing around and got to this point:
> 
> ```
> $SchemaRaw = (Get-Content -raw .\Downloads\test.json)
> $SchemaJson = @($SchemaRaw | ConvertFrom-Json)
> 
> $SchemaJson.properties                                   # For all the possible top layer metadata
> $SchemaJson.definitions.installer.properties             # For all the Installer Metadata
> ```
> 
> Now, if there is a way to clean that up then we have a very nice list, would also be useful for using the patterns and min/max letters instead of hard coding it into the script

Thats about as far as I got too. I also know how to create a function that will take a list of properties and sort an object by those properties. The issue I'm having right now is that the top layer metadata that I get from the json looks like this:
```
@{PackageIdentifier=; PackageVersion=; Channel=; InstallerLocale=; Platform=; MinimumOSVersion=; InstallerType=; Scope=; InstallModes=; InstallerSwitches=; InstallerSuccessCodes=; UpgradeBehavior=; Commands=; Protocols=; FileExtensions=; Dependencies=; PackageFamilyName=; ProductCode=; Capabilities=; RestrictedCapabilities=; Installers=; ManifestType=; ManifestVersion=}
```

Trenly commented:
I think I got it working to pull the order from the schema. Give me a bit and I'll push some changes

Trenly committed:
Sort Yaml Keys

Trenly commented:
@OfficialEsco  Please let me know if that is working on your end

OfficialEsco commented:
> @OfficialEsco Please let me know if that is working on your end

*cough*
```
PS C:\Users\Esco\GitHub\winget-pkgs> .\Tools\YamlCreate.ps1
Error downloading schemas. Please run the script again.
PS C:\Users\Esco\GitHub\winget-pkgs> .\Tools\YamlCreate.ps1
Error downloading schemas. Please run the script again.
PS C:\Users\Esco\GitHub\winget-pkgs> .\Tools\YamlCreate.ps1
Error downloading schemas. Please run the script again.
PS C:\Users\Esco\GitHub\winget-pkgs> .\Tools\YamlCreate.ps1
Error downloading schemas. Please run the script again.
PS C:\Users\Esco\GitHub\winget-pkgs> .\Tools\YamlCreate.ps1
Error downloading schemas. Please run the script again.
PS C:\Users\Esco\GitHub\winget-pkgs> .\Tools\YamlCreate.ps1
Error downloading schemas. Please run the script again.
PS C:\Users\Esco\GitHub\winget-pkgs> .\Tools\YamlCreate.ps1
Error downloading schemas. Please run the script again.
```

Trenly commented:
> > @OfficialEsco Please let me know if that is working on your end
> 
> _cough_
> 
> ```
> PS C:\Users\Esco\GitHub\winget-pkgs> .\Tools\YamlCreate.ps1
> Error downloading schemas. Please run the script again.
> ```
> 
Does it fix it if you change it to this:
```powershell
    $LocaleSchema = @(Invoke-WebRequest 'https://raw.githubusercontent.com/microsoft/winget-cli/master/schemas/JSON/manifests/v1.0.0/manifest.locale.1.0.0.json' -UseBasicParsing| ConvertFrom-Json)
    $LocaleProperties = (ConvertTo-Yaml $LocaleSchema.properties | ConvertFrom-Yaml -Ordered).Keys
    $VersionSchema = @(Invoke-WebRequest 'https://raw.githubusercontent.com/microsoft/winget-cli/master/schemas/JSON/manifests/v1.0.0/manifest.version.1.0.0.json' -UseBasicParsing| ConvertFrom-Json)
    $VersionProperties = (ConvertTo-Yaml $VersionSchema.properties | ConvertFrom-Yaml -Ordered).Keys
    $InstallerSchema = @(Invoke-WebRequest 'https://raw.githubusercontent.com/microsoft/winget-cli/master/schemas/JSON/manifests/v1.0.0/manifest.installer.1.0.0.json' -UseBasicParsing| ConvertFrom-Json)
```

OfficialEsco commented:
Yup that worked

Trenly commented:
I could probably also use the Schema parameters to simplify other portions of the script, but I don't want to introduce any new errors. I'd be happy to fully validate this version (and fix any issues) and merge it in before looking at some of the optional simplifications

OfficialEsco commented:
Do what you feel like you can do and if you want to do it, you've done a great improvement already 😁

Trenly commented:
> Do what you feel like you can do and if you want to do it, you've done a great improvement already 😁

*Minor. It's only a minor improvement.

I'll continue to test this locally and see if I find any other bugs. If you have any issues with it, please do let me know. If nobody finds any in the next week(?) or so, then I think it will be good to merge in. What do you think?

OfficialEsco commented:
Naaaah you just don't know it yet 😁
Should be fine to merge once i get to stress test it, might be some reeeally wide edge cases that don't get picked up but that should be patchable.
Should probably get a reply from @denelon about that License tho

denelon commented:
@Trenly are we looking to include the source, or just use the tool? It looks good on the surface, I just need to make sure whether or not we need to put acknowledgement in our files.

Trenly commented:
> @Trenly are we looking to include the source, or just use the tool? It looks good on the surface, I just need to make sure whether or not we need to put acknowledgement in our files.

Just use the tool

OfficialEsco commented:
Spotted a issue with the New/Update tag in #21984 and #21985, probably because of the sub folder structure

Trenly commented:
> Spotted a issue with the New/Update tag in #21984, #21985 and #21987, probably because of the sub folder structure

Should be fixed now

OfficialEsco commented:
Not sure what happened here, but its not separating the tags
```
[Optional] Enter any tags that would be useful to discover this tool. For example: zip, c++ (Max 16)
Old Variable: vnc remote-management remote-desktop
Tags: vnc, remote-management, remote-desktop
```

Trenly commented:
> Not sure what happened here, but its not separating the tags
> 
> ```
> [Optional] Enter any tags that would be useful to discover this tool. For example: zip, c++ (Max 16)
> Old Variable: vnc remote-management remote-desktop
> Tags: vnc, remote-management, remote-desktop
> ```
> 
> ```
> [Optional] Enter any tags that would be useful to discover this tool. For example: zip, c++ (Max 16)
> Old Variable: system size management disk space available harddisk harddrive
> ```
> 
> However when updating k-lite its okay?
> 
> ```
> [Optional] Enter any tags that would be useful to discover this tool. For example: zip, c++ (Max 16)
> Old Variable: klite, k-lite, klite basic, video, audio, player, mkv, mpeg, mp4, multimedia
> Tags:
> ```
> 
> Just noticed #22072, #22070, #22069 and #22068 missed the `Update:` part
> It also causing the Unexpected file issue 😕

I'll look into the tags issue. That is a bit odd.

As for the Unexpected file, I have no idea what is happening there. It must be something with the way that the automatic PR submissions are being handled, since it's pulling it in as a commit and not a file. Additionally, the commit that is being pulled in happened after my latest commit to this branch. Perhaps I need to rebase on master again?

denelon commented:
We need to add the proper attribution and hook up our component governance interfaces for powershell-yaml in our build pipelines.

denelon commented:
https://github.com/microsoft/winget-pkgs/issues/22081

OfficialEsco commented:
Boi, i think my git part might be screwed up and missing catch issue 😬
#22118
```
Submit PR?
Do you want to submit your PR now?
From https://github.com/microsoft/winget-pkgs
 * branch                master     -> FETCH_HEAD
error: The following untracked working tree files would be overwritten by checkout:
        manifests/u/UnityTechnologies/UnityHub/2.4.5/UnityTechnologies.UnityHub.yaml
Please move or remove them before you switch branches.
Aborting
warning: LF will be replaced by CRLF in manifests/u/UnityTechnologies/UnityHub/2.4.5/UnityTechnologies.UnityHub.locale.en-US.yaml.
The file will have its original line endings in your working directory
warning: LF will be replaced by CRLF in manifests/u/UnityTechnologies/UnityHub/2.4.5/UnityTechnologies.UnityHub.yaml.
The file will have its original line endings in your working directory
[pr/21204 a60cfab25] Update: UnityTechnologies.UnityHub version 2.4.5
 3 files changed, 67 insertions(+)
 create mode 100644 manifests/u/UnityTechnologies/UnityHub/2.4.5/UnityTechnologies.UnityHub.installer.yaml
 create mode 100644 manifests/u/UnityTechnologies/UnityHub/2.4.5/UnityTechnologies.UnityHub.locale.en-US.yaml
 create mode 100644 manifests/u/UnityTechnologies/UnityHub/2.4.5/UnityTechnologies.UnityHub.yaml
fatal: The upstream branch of your current branch does not match
the name of your current branch.  To push to the upstream branch
on the remote, use

    git push github-desktop-trenly HEAD:YamlParsing

To push to the branch of the same name on the remote, use

    git push github-desktop-trenly HEAD

To choose either option permanently, see push.default in 'git help config'.
? Where should we push the 'pr/21204' branch? OfficialEsco/winget-pkgs

Creating pull request for OfficialEsco:pr/21204 into master in microsoft/winget-pkgs

remote:
remote:
Branch 'pr/21204' set up to track remote branch 'pr/21204' from 'origin'.
To https://github.com/OfficialEsco/winget-pkgs.git
 * [new branch]          HEAD -> pr/21204
https://github.com/microsoft/winget-pkgs/pull/22118
```

Trenly commented:
@OfficialEsco  I've reverted my last change in case that was causing that issue. I haven't touched the commits portion other than setting the commit type,  

I'm not 100% sure why that commit [#0c15d79](https://github.com/microsoft/winget-pkgs/pull/22089/commits/0c15d79564aab3f95fe41490a806b831ec0f1799) keeps being pulled in with the new version. If you would be able to provide any help or insight, I'm fairly certain it has something to do with the Git code, but I'm not sure what

OfficialEsco commented:
> @OfficialEsco I've reverted my last change in case that was causing that issue. I haven't touched the commits portion other than setting the commit type,
> 
> I'm not 100% sure why that commit [#0c15d79](https://github.com/microsoft/winget-pkgs/pull/22089/commits/0c15d79564aab3f95fe41490a806b831ec0f1799) keeps being pulled in with the new version. If you would be able to provide any help or insight, I'm fairly certain it has something to do with the Git code, but I'm not sure what

don't think so, but i will have to rework this part anyways, if i can find a smart way to do it..

`git fetch upstream` is supposed to update upstream/master

`git checkout` is supposed to make a branch based on upstream/master, will have to implement `do -> catch error and fatal`

I've been looking into it at times, but i always get stuck cuz its not doing what i want it to do >_>
Can i somehow submit a commit to this PR trough a PR on your repo if i find a solution?

Trenly commented:
> I've been looking into it at times, but i always get stuck cuz its not doing what i want it to do >_>
> Can i somehow submit a commit to this PR trough a PR on your repo if i find a solution?

I'll look into it too, but It may take some time.  
Yeah, my fork should be public, so feel free to submit a PR on my repo and I'll merge it through to the branch if you do find a solution

OfficialEsco committed:
Git error handling

quhxl commented:
Sometimes the package version is surrounded by quotation marks. I believe that this affects compatibility with the `--version` parameter in `winget install`.

OfficialEsco commented:
> Sometimes the `PackageVersion` is surrounded by quotation marks. I believe that this affects compatibility with the `--version` parameter in `winget install`.

Have you tested this? If it affects winget install this should probably be a issue in winget-cli as the expected behaviour should be to ignore quotation marks.

quhxl commented:
> Have you tested this? If it affects winget install this should probably be a issue in winget-cli as the expected behaviour should be to ignore quotation marks.

I remembered that this caused problems. But I could not reproduce it now. Seems to be an error on my part. I guess it's a purely aesthetic issue then.

OfficialEsco committed:
Fixed Moniker prompt for NewLocale

OfficialEsco commented:
> Edit: Maybe it was related to an extra space or something like that back then. But I can't remember anymore.

Mkay, AFAIK quotes are being ignore, however when doing -`-version` you also need to do `--version"13 37"` 
The issue i remember was with a folder structure which required `1337-alpha` instead of `1337 alpha`, might have been 7-Zip related.

Trenly commented:
I did look at the quotes in the version. There are many packages that currently have quotes around the version number and are working just fine with all the winget functions. I haven’t run into any issues in that regard

OfficialEsco commented:
Awaiting merge 🙄 https://github.com/Trenly/winget-pkgs/pull/1

Trenly commented:
@OfficialEsco Did your git changes work? I want to be sure that everything works properly before I rebase it just to pull in the few minor changes that were made on master

OfficialEsco commented:
Yap had no issues for 7 days now.
There is however one tiny issues where moniker gets into the Locale manifest and throws a warning, which my prompt fix did not solve as it finds the Moniker when it reads the old schema 🤦‍♂️

Now that am looking at the code it looks like you tweaked this part

quhxl commented:
I don't know if this is a bug, but when updating [this](https://github.com/microsoft/winget-pkgs/tree/master/manifests/h/Hashicorp/Vagrant) app from version 2.2.17 to 2.2.18 (#22981), the InstallerSwitches from the old manifest were carried over, even though I specified new ones.

It looked something like this:

```
# Created with YamlCreate.ps1 v2.0.0 using YAML parsing
# yaml-language-server: $schema=https://aka.ms/winget-manifest.installer.1.0.0.schema.json

PackageIdentifier: Hashicorp.Vagrant
PackageVersion: 2.2.18
MinimumOSVersion: 10.0.0.0
InstallerType: msi
InstallModes:
- interactive
- silent
- silentWithProgress
InstallerSwitches: 
  Silent: /qn /norestart
  SilentWithProgress: /qn /norestart
InstallerSuccessCodes:
- 3010
- 1706
Commands:
- vagrant
Installers:
- InstallerLocale: en-US
  Architecture: x64
  InstallerType: msi
  Scope: machine
  InstallerUrl: https://releases.hashicorp.com/vagrant/2.2.18/vagrant_2.2.18_x86_64.msi
  InstallerSha256: AED44D72E72D2183354C7E2E0F6118E289BECBA0491A67EB2E5551701B70D359
  InstallerSwitches:
    Custom: /norestart
  UpgradeBehavior: install
  ProductCode: '{4E8CDF07-51A0-4A7D-9F70-984DA870DB37}'
- InstallerLocale: en-US
  Architecture: x86
  InstallerType: msi
  Scope: machine
  InstallerUrl: https://releases.hashicorp.com/vagrant/2.2.18/vagrant_2.2.18_i686.msi
  InstallerSha256: B4F1DA25CD8E536289988DCF922BA3EBD0469FE1C1DE43E2C20C4CE707F491DD
  InstallerSwitches:
    Custom: /norestart
  UpgradeBehavior: install
  ProductCode: '{5439D812-3EA5-4C78-B415-AF196E7B5D0D}'
ManifestType: installer
ManifestVersion: 1.0.0
```

OfficialEsco commented:
It's intended, this is how it catches Metadata fields we have not specified in the script.

However since I made it so the Installer Metadata is hard-coded this miiight cause confusion some places 🙄

I have been thinking about something like this:
```
If (Metadata in Installers are equal) {
  Set global field
} else {
  Clear global field and keep in Installers
}
```
However not sure how to do that

OfficialEsco commented:
Hmm @Trenly i was looking into implementing fetching minItems and maxItems from the Schema last night, everything went smoothly until testing came around
```
[Optional] Enter any Protocols the application provides a handler for. For example: http, https (Max 16)
Protocols: 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20

[Optional] Enter any Commands or aliases to run the application. For example: msedge (Max 16)
Commands: 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20

[Optional] List of additional non-zero installer success exit codes other than known default values by winget (Max 16)
InstallerSuccessCodes: 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20
```

After some head scratching i realized the `-gt` switch actually never worked as expected since i implemented functions..

Currently:    `} while (($FileExtensions -split ", ").Count -gt $MaxItemsFileExtensions)`
Correct:    `} while (($script:FileExtensions -split ", ").Count -gt $MaxItemsFileExtensions)`

```
[Optional] Enter any Protocols the application provides a handler for. For example: http, https (Max 16)
Protocols: 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20

[Optional] Enter any Protocols the application provides a handler for. For example: http, https (Max 16)
Protocols: 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20

[Optional] Enter any Protocols the application provides a handler for. For example: http, https (Max 16)
Protocols: 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16

[Optional] Enter any Commands or aliases to run the application. For example: msedge (Max 16)
Commands: 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20

[Optional] Enter any Commands or aliases to run the application. For example: msedge (Max 16)
Commands: 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16

[Optional] List of additional non-zero installer success exit codes other than known default values by winget (Max 16)
InstallerSuccessCodes: 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20

[Optional] List of additional non-zero installer success exit codes other than known default values by winget (Max 16)
InstallerSuccessCodes: 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16

[Optional] List of supported installer modes. Options: interactive, silent, silentWithProgress
InstallModes: 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20

[Optional] List of supported installer modes. Options: interactive, silent, silentWithProgress
InstallModes: 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16

[Optional] List of supported installer modes. Options: interactive, silent, silentWithProgress
InstallModes: 1, 2, 3
```

OfficialEsco committed:
minItems, maxItems, pattern, variable fix

* Implemented minItems from Schema
* Implemented maxItems from Schema
* Implemented pattern from Schema
* Implemented available Architecture's from Schema
* Implemented available InstallerType's from Schema
* Added $script: to some variables which needs to stay throughout the script

Only tested with one manifest as of now

Cleanup If/Else logic & Rebase on master

vedantmgoyal2009 commented on Tools/YamlCreate.ps1:
```diff
@@ -25,7 +28,39 @@ $ScriptHeader = '# Created using YamlCreate.ps1 v1.1.5'
+    - Handle writing null parameters as comments
+    - Ensure licensing for powershell-yaml is met
+#>
+
+if (Get-Module -ListAvailable -Name powershell-yaml) {
+} 
+else {
+    try {
+        Install-Module -Name powershell-yaml -Force -Repository PSGallery -Scope CurrentUser
+    } 
+    catch {
+        Throw "Unmet dependency. 'powershell-yaml' unable to be installed successfully."
+    }
+}
+
```
```suggestion

if (-not(Get-Module -ListAvailable -Name powershell-yaml)) {
    try {
        Install-Module -Name powershell-yaml -Force -Repository PSGallery -Scope CurrentUser
    } 
    catch {
        Throw "Unmet dependency. 'powershell-yaml' unable to be installed successfully."
    }
}

```

vedantmgoyal2009 replied:
Just a small change...

vedantmgoyal9 committed:
Update method of getting Product Code from installer

vedantmgoyal2009 commented:
I have created a PR https://github.com/Trenly/winget-pkgs/pull/3. If it seems good, please merge.

quhxl commented:
I noticed a small "bug": The tags are not being indented. I don't know if this is intentional, but I assume it is not.

Trenly commented:
> I noticed a small "bug": The tags are not being indented. I don't know if this is intentional, but I assume it is not.

As far as I’m aware, the indentation doesn’t matter for the tags as it is a Yaml string list and is at the top level. It isn’t intentional, I think thats just how powershell-yaml handles it. I’ll look to see if there is any documentation on setting indentation rules though

quhxl commented:
It works fine. I also meant that only as a cosmetic "bug" that is part of the consistency of the manifests.

OfficialEsco commented:
> It works fine and is not mandatory. I also meant that only as a cosmetic "bug" that is part of the consistency of the manifests.

Can you explain to me like I'm 5 years old what you mean by `indented`?
This?
```
Tags:
  - l33t
  - gamer
```
AFAIK Yaml's are supposed to be lame like this
```
Tags:
- l33t
- gamer
```
at least that's what wingetbot and wingetcreate does, trying to make it look similar so when we update the Metadata it doesn't change more then needed

quhxl commented:
Yes that's what I meant, but if it's handled the same way with the other tools then I would leave it as is. For visibility, however, it is better IMHO if "subcategories" are indented. I think this is also the way it is mostly handled for ease of use.

OfficialEsco committed:
More CommitTypes

OfficialEsco committed:
SignatureSha256 and PackageFamilyName (#5)

vedantmgoyal9 committed:
Fill in PR Parameters #21940 (#4)

vedantmgoyal9 committed:
Detect PackageFamilyName for MSIX/APPX (#8)

* Update YamlCreate.ps1

* Update YamlCreate.ps1

quhxl commented:
Is it possible that the new .validation files break the script?

With the file in the directory:

![noteworking](https://user-images.githubusercontent.com/69106310/128614434-196a20a0-b5aa-4703-bc6c-f32fd55831aa.png)

Without the file in the directory:

![working](https://user-images.githubusercontent.com/69106310/128614455-13c0dd22-55c9-43b3-aa21-800df1da1bec.png)

OfficialEsco commented:
Hah, i was just trying to update OneMore too and got the same issue 😁
```
[Required] Enter the Package Identifier, in the following format <Publisher shortname.Application shortname>. For example: Microsoft.Excel
PackageIdentifier: stevencohn.OneMore

[Required] Enter the version. for example: 1.33.7
Version: 3.28
Found Existing Version: OneMore

Get-ChildItem : Cannot find path 'C:\Users\Esco\GitHub\winget-pkgs\manifests\s\stevencohn\OneMore\OneMore' because it
does not exist.
At C:\Users\Esco\GitHub\winget-pkgs\Tools\YamlCreate.ps1:1275 char:36
+ ... cript:OldManifests = Get-ChildItem -Path "$AppFolder\..\$LastVersion"
+                          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : ObjectNotFound: (C:\Users\Esco\...OneMore\OneMore:String) [Get-ChildItem], ItemNotFound
   Exception
    + FullyQualifiedErrorId : PathNotFound,Microsoft.PowerShell.Commands.GetChildItemCommand

Error: Version OneMore does not contain the required manifests
At C:\Users\Esco\GitHub\winget-pkgs\Tools\YamlCreate.ps1:1294 char:9
+         Throw "Error: Version $LastVersion does not contain the requi ...
+         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : OperationStopped: (Error: Version ...uired manifests:String) [], RuntimeException
    + FullyQualifiedErrorId : Error: Version OneMore does not contain the required manifests
```

Works as expected when you indeed delete the .validation file

ItzLevvie commented on Tools/YamlCreate.ps1:
```diff
@@ -307,39 +239,82 @@ Function Read-WinGet-InstallerValues {
+        do {
+            Write-Host
+            Write-Host -ForegroundColor 'Yellow' -Object '[Recommended] Enter the installer PackageFamilyName'
+            Write-Host -ForegroundColor 'White' "[F] Find Automatically [Note: This will install the pacakge to find Family Name and then removes it.]"
```
```suggestion
            Write-Host -ForegroundColor 'White' "[F] Find Automatically [Note: This will install the package to find Family Name and then removes it.]"
```

ItzLevvie replied:
cc @Trenly @vedantmgoyal2009: you missed a typo here.

Trenly commented:
> Is it possible that the new .validation files break the script?
> 
> With the file in the directory:
> 
> ![noteworking](https://user-images.githubusercontent.com/69106310/128614434-196a20a0-b5aa-4703-bc6c-f32fd55831aa.png)
> 
> Without the file in the directory:
> 
> ![working](https://user-images.githubusercontent.com/69106310/128614455-13c0dd22-55c9-43b3-aa21-800df1da1bec.png)

@OfficialEsco This should be as simple as using the exclude switch for the get child items; I can’t get to it this weekend, but if you can’t figure it out I should be able to

OfficialEsco commented:
🤦‍♂️ It was that easy... I couldn't for the life of me figure out how to do it with Where-Object 🤦‍♂️

OfficialEsco committed:
Exclude .validation (#9)

* Exclude .validation

* Added exclude to ExistingVersions

Co-authored-by: Levvie - she/her <11600822+ItzLevvie@users.noreply.github.com>

vedantmgoyal9 committed:
Add new line for PackageFamilyName (#10)

Trenly committed:
Standardize Keypress Menus (#11)

* Standardize Keypress Menus

* Fix: Use variable instead of pipeline

* Fix Help Texts & Spacing

* Cleanup Git Messages

* Filter by Yaml

* Remove debugging line

Co-authored-by: Esco <OfficialEsco@users.noreply.github.com>

Co-authored-by: Kaleb Luedtke <kaleb.carver.luedtke@jci.com>
Co-authored-by: Esco <OfficialEsco@users.noreply.github.com>

OfficialEsco committed:
Replaced PrBodyContentReply `n with array (#12)

Trenly committed:
Menu reduction (#13)

* Simplify Code + Make Issues Entry Safer

* Check issue validity

* Fix unintended Revert

Co-authored-by: Esco <OfficialEsco@users.noreply.github.com>

* Fix unintended Revert

Co-authored-by: Esco <OfficialEsco@users.noreply.github.com>

* Fix unintended Revert

Co-authored-by: Esco <OfficialEsco@users.noreply.github.com>

Co-authored-by: Esco <OfficialEsco@users.noreply.github.com>

OfficialEsco committed:
More patterns (#15)

Added pattern for

* PackageIdentifier
* PackageVersion
* PackageFamilyName
* PackageLocale

Trenly committed:
Readability (#16)

* Add Comparison Functions for Readability

* Simplify Validation Function

* Fix: Missing text

* Remove spaces from split function

* Fix: Resolve renamed variable

* Validate Installer Modes

* Validate File Extensions with Pattern and Length

OfficialEsco commented on Tools/YamlCreate.ps1:
```diff
@@ -220,59 +297,54 @@ Function Read-WinGet-InstallerValues {
+        $script:dest = "$env:TEMP\$FileName"
 
         try {
-            $WebClient.DownloadFile($InstallerUrl, $dest)
+            $WebClient.DownloadFile($InstallerUrl, $script:dest)
         }
         catch {
             Write-Host 'Error downloading file. Please run the script again.' -ForegroundColor Red
             exit 1
         }
```
This part seems to not be working and have for a good while, also we might want to ask for a new URL instead of exiting

Trenly replied:
> This part seems to not be working and have for a good while, also we might want to ask for a new URL instead of exiting
> It does work locally, but not in the script..
As in that part hasn't worked at all in a while? What behavior is it exhibiting? Any way I can reproduce it?

OfficialEsco replied:
```
[Required] Enter the Package Identifier, in the following format <Publisher shortname.Application shortname>. For example: Microsoft.Excel
PackageIdentifier: Element.Element

[Required] Enter the version. for example: 1.33.7
Version: 1.33.7
Found Existing Version: 1.7.34

[Required] Enter the download url to the installer.
Url: https://packages.riot.im/desktop/install/win32/x64/Element%20Setup%201.33.7.exe 


Do you want to save the files to the Temp folder?
  [Y] Yes
  [N] No
  [M] Manually Enter SHA256

Enter Choice (default is 'N'): 

Downloading URL. This will take a while...
Error downloading file. Please run the script again.
Time taken: 0 second(s)
Resolve-Path : Cannot find path 'C:\Users\Esco\AppData\Local\Temp\Element%20Setup%201.33.7.exe' because it does not exist.
At C:\Windows\system32\WindowsPowerShell\v1.0\Modules\Microsoft.PowerShell.Utility\Microsoft.PowerShell.Utility.psm1:110 char:36    
+                 $pathsToProcess += Resolve-Path $Path | Foreach-Objec ...
+                                    ~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : ObjectNotFound: (C:\Users\Esco\...up%201.33.7.exe:String) [Resolve-Path], ItemNotFoundException      
    + FullyQualifiedErrorId : PathNotFound,Microsoft.PowerShell.Commands.ResolvePathCommand
 
Get-AppLockerFileInformation : Cannot find path 'C:\Users\Esco\AppData\Local\Temp\Element%20Setup%201.33.7.exe' because it does 
not exist.
At C:\Users\Esco\GitHub\winget-pkgs\Tools\YamlCreate.ps1:340 char:32
+ ... Information = Get-AppLockerFileInformation -Path $script:dest | Selec ...
+                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidArgument: (C:\Users\Esco\...up%201.33.7.exe:String) [Get-AppLockerFileInformation], ItemNotF  
   oundException
    + FullyQualifiedErrorId : CmdletHelper-FailedResolvingPSPath,Microsoft.Security.ApplicationId.PolicyManagement.Cmdlets.GetAppL  
   ockerFileInformationCmdlet
 
Remove-Item : Cannot find path 'C:\Users\Esco\AppData\Local\Temp\Element%20Setup%201.33.7.exe' because it does not exist.
At C:\Users\Esco\GitHub\winget-pkgs\Tools\YamlCreate.ps1:342 char:314
+ ... le', 'CurrentCultureIgnoreCase'))) { Remove-Item -Path $script:dest }
+                                          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : ObjectNotFound: (C:\Users\Esco\...up%201.33.7.exe:String) [Remove-Item], ItemNotFoundException       
    + FullyQualifiedErrorId : PathNotFound,Microsoft.PowerShell.Commands.RemoveItemCommand
```

Hold on, it does seem to work, just not exit? eitherway i think this might need a rework, to not fully exit the script, something like the PR Url check?

Trenly committed:
Add Custom Error Class (#18)

* Create function for adding list parameters

* Remove RAW writing

* Remove Empty Manifest Folders

* minItems, maxItems, pattern, variable fix

* Implemented minItems from Schema
* Implemented maxItems from Schema
* Implemented pattern from Schema
* Implemented available Architecture's from Schema
* Implemented available InstallerType's from Schema
* Added $script: to some variables which needs to stay throughout the script

* Add Custom Error Class

* Cleanup Class Constructor

* Additional error messages + cleanup

* Continue adding error messages

* Switched Local formatter to OTBS + Static Strings (Where applicable)

* Fix Spacing

* Remove unused variables from rebase

Co-authored-by: Esco <OfficialEsco@users.noreply.github.com>

* Fix: Check for package before do-until

* Finish Locale Errors

* Fix Spacing Issues

* Make Enums Case Sensitive

* Fix Unique Items + Incorrect Variables

Co-authored-by: Esco <OfficialEsco@users.noreply.github.com>

OfficialEsco commented:
Tiny uncommon issue, i was going to clean up the `Gitlab.Gitter.IM` folder and create new manifests in `Gitlab.Gitter` but got this error when trying to create a new manifest in that folder that which does not contain any version folders, only the `IM` folder.

```
Selection: 1

[Required] Enter the Package Identifier, in the following format <Publisher shortname.Application shortname>. For example: 
Microsoft.Excel
PackageIdentifier: Gitlab.Gitter

[Required] Enter the version. for example: 1.33.7
Version: 5.0.1         
Error: Version  does not contain the required manifests
At C:\Users\Esco\GitHub\winget-pkgs\Tools\YamlCreate.ps1:1589 char:9
+         Throw "Error: Version $LastVersion does not contain the requi ...
+         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : OperationStopped: (Error: Version ...uired manifests:String) [], RuntimeException
    + FullyQualifiedErrorId : Error: Version  does not contain the required manifests
```

Trenly committed:
Subfolder fix (#21)

* Fix: Subfolder Exists Erroring

* Return

quhxl commented:
![grafik](https://user-images.githubusercontent.com/69106310/131531601-a257044a-1a39-4d55-af0f-66ab25d02cc6.png)

Just got this error while updating [this ](https://github.com/microsoft/winget-pkgs/blob/ecb8789629b6086edba0986f6dd8572a6c7a1975/manifests/s/szTheory/exifcleaner/3.6.0/szTheory.exifcleaner.yaml)manifest.

jedieaston commented on Tools/YamlCreate.ps1:
```diff
@@ -25,1032 +29,1714 @@ $ScriptHeader = '# Created using YamlCreate.ps1 v1.1.6'
+            $script:LastVersion = Split-Path (Split-Path (Get-ChildItem -Path "$AppFolder\..\" -Recurse -Depth 1 -File -Filter '*.yaml').FullName ) -Leaf | Sort-Object $ToNatural | Select-Object -Last 1
+            $script:ExistingVersions = Split-Path (Split-Path (Get-ChildItem -Path "$AppFolder\..\" -Recurse -Depth 1 -File -Filter '*.yaml').FullName ) -Leaf | Sort-Object $ToNatural | Select-Object -Unique
+            Write-Host -ForegroundColor 'DarkYellow' -Object "Found Existing Version: $LastVersion"
+            $script:OldManifests = Get-ChildItem -Path "$AppFolder\..\$LastVersion"
```
```suggestion
            $script:OldManifests = Get-ChildItem -Path "$AppFolder\..\$LastVersion" -Directory
```
To resolve #27518.

vedantmgoyal2009 replied:
I think this PR will take time to merge so we should open another PR for the time being so the users using YamlCreate.ps1 v1.1.6 wouldn't run into any issues.

vedantmgoyal2009 replied:
#27522

jedieaston reviewed (changes requested):
one quick thing.

Trenly committed:
Write Unused Keys as Comments (#22)

* Write Unused Keys as Comments

* Exclude certain keys from appearing as comments

* Comment only Locale + Version

vedantmgoyal9 committed:
Quick Update, Function Reduction, Detect Installer Types, Detect Installer Architectures, and Comments (#27)

* Update YamlCreate.ps1

* auto-detect installer architecture

* Fix https://github.com/Trenly/winget-pkgs/pull/27#issuecomment-921619221

* Fix auto-detection

* Apply suggestions from code review

Co-authored-by: Kaleb Luedtke <trenlymc@gmail.com>

* Move Test + Submit to main

* Update Tools/YamlCreate.ps1

* fix quick-update check

* Reduce input required for quick version update (#4)

* Reduce input required for quick version update

* Fix Product Code

* Don't add null product code if no product code key exists

* improve quick update

* Move prompt outside of Read-Info for clarity

* Move Update Prompt outside of Read-Info for clarity

* Update Messages

* if there are any errors due to this commit, i will revert changes

* Add Comments for code clarity

Co-authored-by: Kaleb Luedtke <trenlymc@gmail.com>

Trenly committed:
Fix error from when functions were moved

Trenly committed:
Handle Manifest Level Parameters (#34)

* Begin Handling Manifest Level Parameters

* Fix InstallModes

Trenly commented:
> ![grafik](https://user-images.githubusercontent.com/69106310/131531601-a257044a-1a39-4d55-af0f-66ab25d02cc6.png)
> 
> Just got this error while updating [this ](https://github.com/microsoft/winget-pkgs/blob/ecb8789629b6086edba0986f6dd8572a6c7a1975/manifests/s/szTheory/exifcleaner/3.6.0/szTheory.exifcleaner.yaml)manifest.

Should be resolved with https://github.com/Trenly/winget-pkgs/commit/a20641f324534a57f8d84534a1c2c2d0c3703b00

Trenly committed:
Add delete manifest functionality (#39)

* Add functionality to manually delete manifests

Trenly committed:
Add settings file and script documentation (#38)

* Do not default installer locale

* Add a script settings file with documentation

* Add setting to suppress quick update warning

* Remove accidental file inclusion

Trenly committed:
Fix: Update SignatureSha256 in quick update mode (#41)

Trenly committed:
Change Settings to allow for negative suppression (#46)

* Change Settings to allow for negative suppression

* Fix sample settings file

* Invalid SandboxTest variable

Co-authored-by: Vedant Mohan Goyal <83997633+vedantmgoyal2009@users.noreply.github.com>

Co-authored-by: Kaleb Luedtke <kaleb.carver.luedtke@jci.com>
Co-authored-by: Esco <OfficialEsco@users.noreply.github.com>
Co-authored-by: Vedant Mohan Goyal <83997633+vedantmgoyal2009@users.noreply.github.com>

vedantmgoyal9 committed:
Cleaner references to linked issues (#49)

Trenly committed:
Command Line Arguments (#50)

* Add switch for settings

* Allow PackageIdentifier and PackageVersion to be passed as optional parameters

* Add -help switch and documentation

* Settings File Location

* Fix spacing issue [Trenly/winget-pkgs#49] Trenly/winget-pkgs#51

Co-authored-by: Kaleb Luedtke <kaleb.carver.luedtke@jci.com>
Co-authored-by: Vedant Mohan Goyal <83997633+vedantmgoyal2009@users.noreply.github.com>

Trenly committed:
Fix Parameter Reading and Condensing(#54)

* Fix: Allow InstallerSwitch keys to be split between installer level and manifest level on a per-key basis

* Fix: Save parameters to variables before removing them from manifest level

Trenly committed:
Fix: Sandboxtest not working

Trenly committed:
Fix: Prompt for PFN if file does not exist (#55)

Trenly committed:
Auto Mode (#58)

* Auto Mode

* Fix for same version

* Keep ProductCodes for .exe files

* Keep ProductCodes for .exe files in Option-2

Trenly committed:
Simplify function names, add debug info (#62)

jedieaston commented on Tools/YamlCreate.ps1:
```diff
@@ -310,7 +310,7 @@ Function Request-Installer-Url {
 # Sets the $script:Installers value as an output
 # Returns void
-Function Read-WinGet-InstallerValues {
+Function Read-Installer-Values {
```
Isn't two dashes in a PowerShell function name against the law?

Maybe this should be changed to `Read-InstallerValues`. 

(Sorry I didn't see your initial PR.)

Trenly replied:
I don't know why it would be against the law? The [current version](https://github.com/microsoft/winget-pkgs/blob/a51d74b195208f815909a86f2aed5dd75fbc46a6/Tools/YamlCreate.ps1#L84) of the script has 2 dashes in function names. It seems to work just fine. Its not difficult to change though if we want to

jedieaston replied:
I can't find a citation, but I think Microsoft recommends using a single dash as above to stick to the `Verb-Noun` pattern they started. 

Since these functions aren't for external use outside the script anyway, I guess it doesn't matter. just something I randomly noticed.

Trenly committed:
Fix: Allow autoupdate to update old package versions (#66)

Trenly committed:
Make Simple Update automatically detect Sha256, SignatureSha256, and ProductCode without prompts (#61)

* Make Simple Update automatically detect Sha256, SignatureSha256, and ProductCode without prompts

* Add Parameter Mode (#64)

* Add Parameter Mode

* Update YamlCreate.ps1

* Update YamlCreate.md

Co-authored-by: Vedant Mohan Goyal <83997633+vedantmgoyal2009@users.noreply.github.com>

Trenly committed:
Use Unique Branch Names (#67)

* Use Unique Branch Names

* Process Branch Names Safely

Trenly committed:
Fix: Allow values other than en-US for default locale (#70)

jedieaston committed:
Made it so git config was only modified for the local repo. (#72)

* Only modify git config for local repo.

* Fixed 2214.

vedantmgoyal9 committed:
Support for settings on Linux and macOS (#73)

* Support for settings on Linux and macOS

* Easton's suggestion

* Update YamlCreate.md

* Update Tools/YamlCreate.ps1

Co-authored-by: Kaleb Luedtke <trenlymc@gmail.com>

* Update YamlCreate.ps1

Co-authored-by: Kaleb Luedtke <trenlymc@gmail.com>

Trenly committed:
Check for open PR's before submitting (#69)

* Check for open PR's before submitting

* Rebase on c82b39f

* Use API instead of CLI

* Exit when user-choice to terminate

Co-authored-by: Vedant Mohan Goyal <83997633+vedantmgoyal2009@users.noreply.github.com>

Co-authored-by: Vedant Mohan Goyal <83997633+vedantmgoyal2009@users.noreply.github.com>

Trenly committed:
Throw when script error (#74)

Trenly committed:
Fix: Add locale when converting from singleton (#75)

Trenly committed:
Fix: Branch names when deleting (#79)

Trenly committed:
Fix: Don't check for PRs on deletion (#80)

Trenly committed:
Update PR Content when removing a manifest (#82)

* Update PR Content when removing a manifest

* More realistic character count limits

vedantmgoyal9 committed:
Add is:pr to exclude issues from results (#88)

Trenly committed:
Ensure file names are valid before saving (#85)

* Ensure file names are valid before saving

* Fix: Catch when content disposition doesn't exist

* Fix UserAgent not following redirects

* Additional Web Request Parameters

* Fix: Min descriptor length + Agent

Trenly committed:
Fix: Escape Regex in variables (#91)

Trenly committed:
Fix: Only add moniker to defaultLocale (#92)

Trenly committed:
Chore: Add references to YamlCreate documentation (#93)

vedantmgoyal9 committed:
Typo (#94)

Trenly commented:
> @Trenly I just merged the component governance portion in. We're good to go 😊

You're the best. I think the script is in a stable spot now too. Its been mostly minor fixes for the past week

@jedieaston @OfficialEsco @vedantmgoyal2009  - Would you mind doing a final review?

Trenly committed:
Chore: Add note to enable settings

Co-authored-by: denelon <denelon@microsoft.com>

denelon reviewed (commented):
A minor addition to README.md.

This is a large change, so we will take a deeper look early next week. It looks like a great set of enhancements!

Trenly committed:
Chore: Remove To-do message

vedantmgoyal2009 commented:
Should we give it a go with PSScriptAnalyser?

vedantmgoyal9 committed:
Update YamlCreate.ps1

jedieaston commented:
![image](https://user-images.githubusercontent.com/21368066/137569733-e78a2df7-b885-458b-817d-c0f6d5c9da76.png)

I'm guessing a lot of this is noise.

Trenly commented:
> ![image](https://user-images.githubusercontent.com/21368066/137569733-e78a2df7-b885-458b-817d-c0f6d5c9da76.png)
> 
> I'm guessing a lot of this is noise.

I think so, but we can still clean some of it up

vedantmgoyal2009 commented:
> ![image](https://user-images.githubusercontent.com/21368066/137569733-e78a2df7-b885-458b-817d-c0f6d5c9da76.png)
> 
> I'm guessing a lot of this is noise.

I've been fixing this. Will open a PR soon. Also, should `Write-Host`s be replaced by `Write-Output`s?

jedieaston commented:
> Also, should Write-Hosts be replaced by Write-Outputs?
Yes please.

Trenly committed:
Suppress Write-Host warnings from ScriptAnalyzer

Trenly committed:
Rename Write-Locale-Manifests -> Write-LocaleManifest

Trenly committed:
Rename Write-Version-Manifest -> Write-VersionManifest

Trenly committed:
Rename Write-Installer-Manifest -> Write-InstallerManifest

Trenly committed:
Rename Read-WinGet-LocaleManifest -> Read-LocaleMetadata

Trenly committed:
Rename Read-WinGet-InstallerManifest -> Read-InstallerMetadata

Trenly committed:
Rename Read-Installer-Values-Minimal -> Read-QuickInstallerEntry

Trenly committed:
Rename Read-Installer-Values -> Read-InstallerEntry

Trenly committed:
Rename Enter-PR-Parameters -> Read-PRBody

Trenly committed:
Rename Write-Colors -> Write-MulticolorLine

Trenly committed:
Move integer validation (#99)

Trenly committed:
Rename String.Validate -> Test-String (#101)

Trenly committed:
Move Downloading of installer to a separate function

Trenly committed:
Use Named Parameters for AddYamlParameter

Trenly committed:
Use Named Parameters for AddYamlListParameter

Trenly committed:
Fix missing constructor parameter

Trenly committed:
Use Named Parameters for PromptInstallerManifestValue

Trenly committed:
Support ShouldProcess for removing manifests

Trenly commented:
@jedieaston @vedantmgoyal2009 I believe these are the only issues remaining:
```raw
RuleName                            Severity     ScriptName Line  Message                                                     
--------                            --------     ---------- ----  -------                                                     
PSReviewUnusedParameter             Warning      YamlCreate 886   The parameter 'InputObject' has been declared but not used. 
                                                 .ps1                                                                         
PSReviewUnusedParameter             Warning      YamlCreate 889   The parameter 'NoComments' has been declared but not used.  
                                                 .ps1                                                                         
```

I looked and I see these parameters being used, so I think they are false positives

Trenly committed:
More function name updates

Trenly commented:
@denelon I've left the commit history mostly in-tact up until this point; Would you have any objection if I rewrite the history of this branch to reduce the amount of clutter?

Trenly committed:
Fix: Download Method (#107)

Trenly committed:
Check for version when setting proxy (#109)

Trenly commented:
@jedieaston @OfficialEsco Would you mind reviewing this PR for @denelon ?

jedieaston reviewed (dismissed):
I ran through it one more time, there doesn't seem to be any outstanding uh-ohs that we haven't noted.

:shipit:

Trenly committed:
Use defaultLocale due to microsoft/winget-cli#1646 resolving microsoft/winget-cli#1595

Trenly commented:
> There doesn't seem to be any outstanding uh-ohs that we haven't noted.
Found one from the latest fixes to the CLI; Patched it
